File: HouseParty/HouseParty/Models/Posting.swift
//
//  Posting.swift
//  HouseParty
//
//  Created by Keith Okuna on 10/25/21.
//

import Foundation

struct Posting: Codable, Identifiable {
    var id: String? { _id }
    var _id: String?
    var author: PostingAuthor = PostingAuthor()
    var body: String = ""
    var title: String = ""
    var date: String = ""
    var replies: [PostingReply] = []
    var actions: [PostingAction] = []
}

struct PostingAuthor: Codable, Identifiable {
    var id: String? { _id }
    var _id: String?
    var name: String = ""
    var email: String = ""
    var bio: String = ""
}

struct PostingReply: Codable, Identifiable {
    var id: String? { _id }
    var _id: String?
    var author: PostingAuthor = PostingAuthor()
    var action: String = ""
}

struct PostingAction: Codable, Identifiable {
    var id: String? { _id }
    var _id: String?
    var author: PostingAuthor = PostingAuthor()
    var body: String = ""
    var title: String = ""
}

File: server/models/posting.ts
import { ObjectId } from "mongodb";
import { Schema, model, connect, SchemaDefinition} from "mongoose";
import { User, UserModel } from "./user"

import mongoose from 'mongoose';

interface Posting {
    _id: ObjectId;
    body: string;
    title: string;
    date: Date;
    author: User;
    replies:[PostingReply?]
    actions: [PostingAction?]
}

interface PostingAction {
    author: User;
    action: string;
}

interface PostingReply {
    author: User;
    title: string;
    body: string;
    date: string;
}


const postingSchema = new Schema<Posting>({
    body: {
        type: String,
        required: true
    },
    title: {
        type: String,
        required: true
    },
    date: {
        type: Date,
        default: new Date()
    },
    author: {
    },
    replies: [],
    actions: []
});

const PostingModel = model<Posting>('Posting', postingSchema);

export default PostingModel;